graylog2-web-interface/src/components/configurations/SearchesConfig.jsx

```diff
@@ -0,0 +1,288 @@
+  },
+
+  _validationState() {
+    if (this._isValidPeriod()) {
+      return undefined;
+    } else {
+      return 'error';
+    }
+  },
+
+  _formatDuration(duration) {
+    return this._isValidPeriod() ? moment.duration(duration).humanize() : 'invalid';
+  },
+
+  _onPeriodUpdate(field) {
+    return () => {
+      const update = this.state.config;
+      let period = this.refs[field].getValue().toUpperCase();
+
+      if (!period.startsWith('P')) {
+        period = `P${period}`;
+      }
+
+      update[field] = period;
+
+      this.setState({config: update});
+    };
+  },
+
+  _onChecked() {
+    const config = this.state.config;
+
+    if (this.state.limitEnabled) {
+      // If currently enabled, disable by setting the limit to 0 seconds.
+      config.query_time_range_limit = 'PT0S';
+    } else {
+      // If currently not enabled, set a default of 30 days.
+      config.query_time_range_limit = 'P30D';
+    }
+
+    this.setState({config: config, limitEnabled: !this.state.limitEnabled});
+  },
+
+  _onTimerangeOptionRemove(removedIdx) {
+    return () => {
+      const options = this.state.timerangeOptionsList;
+
+      // Remove element at index
+      options.splice(removedIdx, 1);
+
+      this._updateTimerangeOptionsState(options);
+    };
+  },
+
+  _onTimerangeOptionAdd() {
+    const options = this.state.timerangeOptionsList;
+
+    if (options) {
+      options.push({period: 'PT1S', description: ''});
+      this._updateTimerangeOptionsState(options);
+    }
+  },
+
+  _onTimerangeOptionChange(changedIdx, field) {
+    return (e) => {
+      const options = this.state.timerangeOptionsList;
+
+      options.forEach((o, idx) => {
+        if (idx === changedIdx) {
+          let value = e.target.value;
+
+          if (field === 'period') {
+            value = value.toUpperCase();
+            if (!value.startsWith('P')) {
+              value = `P${value}`;
+            }
+          }
+
+          options[idx][field] = value;
+        }
+      });
+
+      this._updateTimerangeOptionsState(options);
+    };
+  },
+
+  _isEnabled() {
+    return this.state.limitEnabled;
+  },
+
+  _saveConfig() {
+    this.props.updateConfig(this.state.config).then(() => {
+      this._closeModal();
+    });
+  },
+
+  _resetConfig() {
+    // Reset to initial state when the modal is closed without saving.
+    this.setState(this.getInitialState());
+  },
+
+  _openModal() {
+    this.refs.searchesConfigModal.open();
+  },
+
+  _closeModal() {
+    this.refs.searchesConfigModal.close();
+  },
+
+  render() {
+    const config = this.state.config;
+    const duration = moment.duration(config.query_time_range_limit);
+    const limit = this._isEnabled() ? `${config.query_time_range_limit} (${duration.format()})` : 'disabled';
+
+    let timerangeOptions = null;
+    let timerangeOptionsSummary = null;
+    if (this.state.timerangeOptionsList) {
+      timerangeOptions = this.state.timerangeOptionsList.map((option, idx) => {
```

Comment:
I like to put these large blocks creating some jsx into a function for clarity, also keeps the render method short. That's something of personal taste, of course :)